glean/rts/ondemand.h

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#pragma once
+
+#include <atomic>
+#include <memory>
+
+namespace facebook {
+namespace glean {
+namespace rts {
+
+/// Values which are created once when needed. OnDemand is thread safe and
+/// always returns the same value. However, it might speculatively create (and
+/// then destroy) additional values.
+template<typename T>
+class OnDemand final {
+public:
+  /// Create an empty OnDemand
+  OnDemand() noexcept : ptr_(nullptr) {}
+
+  OnDemand(OnDemand&& other) noexcept
+    : ptr_(other.ptr_.exchange(nullptr, std::memory_order_acq_rel))
+    {}
+
+  OnDemand& operator=(OnDemand&& other) noexcept {
+    auto p = other.ptr_.exchange(nullptr, std::memory_order_acq_rel);
+    p = ptr_.exchange(ptr_, std::memory_order_release);
```

Comment:
Infer complained here about the first assignment to `p` being unused. Should this be
```suggestion
    p = ptr_.exchange(p, std::memory_order_release);
```